qemu/tests/migration_with_block.py

```diff
@@ -0,0 +1,233 @@
+import logging
+import ast
+import re
+
+from virttest import error_context
+from virttest import utils_test
+from virttest import utils_disk
+from virttest import utils_misc
+
+from avocado.utils import process
+
+from provider.storage_benchmark import generate_instance
+from provider.cdrom import QMPEventCheckCDEject, QMPEventCheckCDChange
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Test block devices with migration..
+
+        Scenario "with_scsi_on2off":
+            1) Boot guest with scsi=on on src host.
+            2) Boot guest with scsi=off on dst host.
+            3) Do live Migration.
+
+        Scenario "with_change_cdrom":
+            1) Run qemu with specified cdrom loaded.
+            2) Check the cdrom info by qmp.
+            3) Check the cdrom's size inside guest.
+            4) Eject cdrom, and check the info again.
+            5) Load a new cdrom image, and check the cdrom info again.
+            6) Check the cdrom's size inside guest.
+            7) Start dest vm with new iso file in listening mode.
+            8) Migrate from src to dst.
+            9) Do system_reset in dst vm.
+
+        Scenario "with_dataplane_on2off":
+            1) Start VM with dataplane (both system disk and data disk).
+            2) For Windows: check whether viostor.sys verifier enabled in guest.
+            3) Do live migration.
+            4) Do iozone testing after migration.
+            5) Reboot guest.
+
+    :param test:   QEMU test object.
+    :param params: Dictionary with the test parameters.
+    :param env:    Dictionary with test environment.
+    """
+    def _get_data_disks_linux():
+        """ Get the data disks by serial or wwn options in linux. """
+        for data_image in params['images'].split()[1:]:
+            extra_params = params.get("blk_extra_params_%s" % data_image, '')
+            match = re.search(r"(serial|wwn)=(\w+)", extra_params, re.M)
+            if match:
+                drive_id = match.group(2)
+            else:
+                continue
+            drive_path = utils_misc.get_linux_drive_path(session, drive_id)
+            if not drive_path:
+                test.error("Failed to get '%s' drive path" % data_image)
+            yield drive_path[5:], params.object_params(data_image)['image_size']
+
+    def _get_data_disks_win():
+        """ Get the data disks in windows. """
+        for data_image in params['images'].split()[1:]:
+            size = params.object_params(data_image)['image_size']
+            yield utils_disk.get_windows_disks_index(session, size)[0], size
+
+    def get_data_disks():
+        """ Get the data disks. """
+        _get_disks = _get_data_disks_win if windows else _get_data_disks_linux
+        for disk, size in _get_disks():
+            yield disk, size
+
+    def format_data_disks():
```

Comment:
maybe the get_data_disks() and format_data_disks() function can be merged to one function? Because the get_data_disks() function is not used elsewhere.

Reply:
The get_data_disks is used in format_data_disks in order to readable, a function is not only for redundancy also wrap the logical blocks readable so that the other develops could know of it more clear.